ksqldb-engine/src/main/java/io/confluent/ksql/utilization/PersistentQuerySaturationMetrics.java

```diff
@@ -0,0 +1,318 @@
+  private static final String STREAMS_THREAD_METRICS_GROUP = "stream-thread-metrics";
+  private static final String THREAD_ID = "thread-id";
+  private static final String QUERY_ID = "query-id";
+
+  private final Map<String, KafkaStreamsSaturation> perKafkaStreamsStats = new HashMap<>();
+  private final KsqlEngine engine;
+  private final Supplier<Instant> time;
+  private final MetricsReporter reporter;
+  private final Duration window;
+  private final Duration sampleMargin;
+
+  public PersistentQuerySaturationMetrics(
+      final KsqlEngine engine,
+      final MetricsReporter reporter,
+      final Duration window,
+      final Duration sampleMargin
+  ) {
+    this(Instant::now, engine, reporter, window, sampleMargin);
+  }
+
+  @VisibleForTesting
+  PersistentQuerySaturationMetrics(
+      final Supplier<Instant> time,
+      final KsqlEngine engine,
+      final MetricsReporter reporter,
+      final Duration window,
+      final Duration sampleMargin
+  ) {
+    this.time =  Objects.requireNonNull(time, "time");
+    this.engine = Objects.requireNonNull(engine, "engine");
+    this.reporter = Objects.requireNonNull(reporter, "reporter");
+    this.window = Objects.requireNonNull(window, "window");
+    this.sampleMargin = Objects.requireNonNull(sampleMargin, "sampleMargin");
+  }
+
+  @Override
+  public void run() {
+    final Instant now = time.get();
+
+    try {
+      final Collection<PersistentQueryMetadata> queries = engine.getPersistentQueries();
+      final Optional<Double> saturation = queries.stream()
+          .map(q -> measure(now, q))
+          .max(PersistentQuerySaturationMetrics::compareSaturation)
+          .orElse(Optional.of(0.0));
+      saturation.ifPresent(s -> report(now, s));
+
+      final Set<String> appIds = queries.stream()
+          .map(PersistentQueryMetadata::getQueryApplicationId)
+          .collect(Collectors.toSet());
+      for (final String appId
+          : Sets.difference(new HashSet<>(perKafkaStreamsStats.keySet()), appIds)) {
+        perKafkaStreamsStats.get(appId).cleanup(reporter);
+        perKafkaStreamsStats.remove(appId);
+      }
+    } catch (final RuntimeException e) {
+      LOGGER.error("Error collecting saturation", e);
+      throw e;
+    }
+  }
+
+  private static int compareSaturation(final Optional<Double> a, final Optional<Double> b) {
+    // A missing value means we could not compute saturation for some unit (thread, runtime, etc)
+    // therefore, the result of the comparison is unknown.
+    if (!a.isPresent()) {
+      return 1;
+    } else if (!b.isPresent()) {
+      return -1;
+    } else {
+      return Double.compare(a.get(), b.get());
+    }
+  }
+
+  private Optional<Double> measure(final Instant now, final PersistentQueryMetadata queryMetadata) {
+    final KafkaStreamsSaturation ksSaturation = perKafkaStreamsStats.computeIfAbsent(
+        queryMetadata.getQueryApplicationId(),
```

Comment:
ditto here, or filter for duplicates if you are intending to have it be per runtime not per query